river/cluster/evostream.py

```diff
@@ -0,0 +1,424 @@
+    ----------
+    clusters
+        A set of final clusters of type `evoStreamMicroCluster`, i.e clusters with structure
+        `(center, last_update, weight)`. These clusters are formed by choosing the cluster solution
+        with highest fitness.
+    centers
+        Final clusters' centers.
+
+    References
+    ----------
+    [^1]: Carnein, M., Trautmann, H. (2018). evoStream – Evolutionary Stream Clustering Utilizing Idle Times.
+          Big Data Research, 14, 101-111. DOI: 10.1016/j.bdr.2018.05.005
+    [^2]: Maulik, U., Bandyopadhyay, S. (2000). Genetic algorithm-based clustering technique.
+          Pattern Recognition, 33(9), 1455-1465. DOI: 10.1016/s0031-3203(99)00137-5
+    [^3]: Hahsler, M., Bolanos, M. Clsutering Data Streams Based on Shared Density between Micro-Clusters,
+          IEEE Transactions on Knowledge and Data Engineering 28(6), 2016, 1449-1461.
+          In Proceedings of the Sixth SIAM International Conference on Data Mining,
+          April 20–22, 2006, Bethesda, MD, USA.
+
+    Examples
+    ----------
+    >>> from river import cluster
+    >>> from river import stream
+
+    >>> X = [
+    ...     [1, 0.5], [1, 0.625], [1, 0.75], [1, 1], [1, 1.125], [1, 1.25],
+    ...     [1, 1.5], [1, 1.75], [1, 2], [4, 1.25], [4, 1.5], [4, 2.25],
+    ...     [4, 2.5], [4, 3], [4, 3.25], [4, 3.5], [4, 3.75], [4, 4],
+    ... ]
+
+    >>> evostream = cluster.evoStream(radius=0.25,
+    ...                               cleanup_interval=6,
+    ...                               evolution_interval=7,
+    ...                               n_clusters=2)
+
+    >>> for x, _ in stream.iter_array(X):
+    ...     evostream = evostream.learn_one(x)
+
+    >>> evostream.predict_one({0: 0, 1: 0}) != evostream.predict_one({0: 4, 1: 3})
+    True
+
+    """
+
+    def __init__(
+        self,
+        radius: float = 1.0,
+        decay_rate: float = 0.01,
+        cleanup_interval: int = 6,
+        evolution_interval: int = 4,
+        initialization_threshold: int = 3,
+        mutation_probability: float = 0.5,
+        population_size: int = 4,
+        n_clusters: int = 2,
+    ):
+        super().__init__()
+        self.time_stamp = 0
+
+        self.radius = radius
+        self.decay_rate = decay_rate
+        self.cleanup_interval = cleanup_interval
+        self.evolution_interval = evolution_interval
+        self.initialization_threshold = initialization_threshold
+        self.mutation_probability = mutation_probability
+        self.population_size = population_size
+        self.n_clusters = n_clusters
+        self.initialized = False
+
+        self.clusters = {}
+        self.centers = {}
+
+        self._evolution_batch = {}
+        self._micro_clusters = {}
+        self._cluster_population = {}
+        self._fitness_cluster_population = {}
+
+    @staticmethod
+    def _binary_crossover(cluster_sol_1, cluster_sol_2):
+        dim = len(cluster_sol_1[0].center)
+        cutoff = math.floor(dim / 2)
```

Comment:
It is not clear to me from the paper whether the cut-off point is "always fixed". Keep in mind that I only skimmed the paper, though. Is the mid-point always used or the cut-off is also randomly defined?

Reply:
You are right, the cut-off point is also randomly defined. Changed to `cutoff = random.randint(dim - 1)`.